File: dynamic_listings/C++/default/AI/jump_shoot.cpp
#include <functional>
#include <cstdlib>

class Player{
    // ...
    public:
        void registerShootingObserver(std::function callback){
            // Function used to register an observer that will be called when the
            // player shoots a projectile.
        }
};

class JumpingBoss{
    // A boss that jumps when the player shoots. Sometimes.

    private:
        bool player_shot = false;
        bool on_ground = false;
        float y_velocity = 0.0;

    public:

        JumpingBoss(float x, float y, Player player){
            // ...
            player.registerShootingObserver(this.setPlayerShot);
            // ...
        }

        void setPlayerShot(){
            // Sets a state that tells the AI that the player shot a bullet
            player_shot = True;
        }

        void jump(){
            // Sets the boss velocity to -10, making it jump
            if (on_ground){
                y_velocity = -10;
            }
        }

        void update(float dt){
            // ...
            if (player_shot){
                if (std::rand() % 5 + 1 == 1){
                    // Jump only 20% of the times the player shoots
                    jump();
                }
            }
            // We reset player_shot to false, if we didn't the boss would jump
            // a lot more often than 20% of the time
            player_shot = False;
            // ...
        }
};

File: dynamic_listings/rust/default/AI/jump_shoot.rs
# //! ```cargo
# //! [dependencies]
# //! rand = "0.10"
# //! ```
#
# #![allow(dead_code, unused)]
struct Player {
    // ...
}

impl Player {
    fn register_shooting_observer(&self, callback: fn(&mut JumpingBoss)) {
        // Function used to register an observer that will be called when the
        // player shoots a projectile.
    }
}

struct JumpingBoss {
    player_shot: bool,
    on_ground: bool,
    y_velocity: f32,
}

impl JumpingBoss {
    pub fn new(x: f32, y: f32, player: &Player) {
        player.register_shooting_observer(Self::set_player_shot);
    }

    pub fn set_player_shot(&mut self) {
        self.player_shot = true;
    }

    pub fn jump(&mut self) {
        if self.on_ground {
            self.y_velocity = -10.;
        }
    }

    pub fn update(&mut self, dt: f32) {
        // ...
        if self.player_shot {
            if rand::random_range(1..=5) == 1 {
                // Jump only 20% of the times the player shoots
                self.jump();
            }
        }
        // We reset player_shot to false, if we didn't the boss would jump
        // a lot more often than 20% of the time
        self.player_shot = false;
        // ...
    }
}
